data/stiles_data/parse_la_data.py

```diff
@@ -0,0 +1,404 @@
+from pathlib import Path
+
+import geopandas as gpd
+
+
+class CityParser(object):
+
+    def __init__(self, path: Path):
+        geo_jsons = [p for p in path.iterdir() if p.is_file() and p.suffix == '.geojson']
+        self.city = path.parts[-1]
+        assert len(geo_jsons) <= 1
+        if len(geo_jsons) > 0:
+            self.geo_json_path = geo_jsons[-1]
+        else:
+            self.geo_json_path = None
+
+    def get_maximal_df(self):
+        assert self.geo_json_path
+        df = gpd.read_file(str(self.geo_json_path.absolute())).assign(city=self.city)
+        return self.lat_lon_from_geometry(df)
+
+    @staticmethod
+    def lat_lon_from_geometry(df, y_is_lat=True):
+        if y_is_lat:
+            return df.assign(
+                latitude=df['geometry'].apply(lambda p: p.y),
+                longitude=df['geometry'].apply(lambda p: p.x)
+            )
+        return df.assign(
+            latitude=df['geometry'].apply(lambda p: p.x),
+            longitude=df['geometry'].apply(lambda p: p.y)
+        )
+
+    @staticmethod
+    def cat_parser(df, min_field, max_field, og_field, cats):
+        actual_cats = [cat for cat in df[og_field].unique().tolist() if set(cat.strip()) != {'-'}]
+        if len(actual_cats) > len(cats):
+            raise RuntimeError(f'{len(cats)} categories but categories in df={df[og_field].unique().tolist()}')
+        df[min_field] = -1
+        df[max_field] = -1
+        for cat in cats:
+            mask = df[og_field].str.strip() == cat
+            if len(cat.split('-')) == 2:
+                min_val, max_val = cat.split('-')
+                df.loc[mask, min_field] = int(min_val)
+                df.loc[mask, max_field] = int(max_val)
+            elif cat.endswith('+'):
+                min_val = int(cat[:-1])
+                df.loc[mask, min_field] = int(min_val)
+            elif cat.startswith('>'):
+                min_val = int(cat[1:])
+                df.loc[mask, min_field] = int(min_val)
+
+        return df
+
+
+class LosAngelesCityParser(CityParser):
+    def __init__(self, path: Path):
+        super().__init__(path)
+
+    def get_maximal_df(self):
+        df = super().get_maximal_df()
+        df = df.assign(name_common=df['species'].str.title())
+        df = self.lat_lon_from_geometry(df)
+        return df[['name_common', 'latitude', 'longitude', 'city']]
+
+
+class LosAngelesCountyParser(CityParser):
+
+    def __init__(self, path: Path):
+        super().__init__(path)
+
+    def get_maximal_df(self):
+        df = super().get_maximal_df()
+        df = df.assign(
+            name_common=df['SPECIES'].str.title(),
+            diameter_min_in=df['DIAMETER']
+        )
+        return df[['name_common', 'latitude', 'longitude', 'diameter_min_in', 'city']]
+
+
+class AgouraHillsParser(CityParser):
+
+    def __init__(self, path: Path):
+        super().__init__(path)
+
+    def get_maximal_df(self):
+        df = super().get_maximal_df()
+        df = df.assign(
```

Comment:
question from @daniellezegelstein @Cdower @cajaks2 

why is the variable df reassigned in every function? why isn't the variable renamed? @allentran

Reply:
There is a parent class `CityParser` that does some useful things common to all cities.

Each cities actual parser is a subclass of `CityParser`, e.g `BeverlyHillsParser`. Each subclass reimplements/overloads (I forget the correct term) the method `get_maximal_df()` to do all the custom things required for that city. The name `df` refers to dataframe and each operation is just adding/changing stuff to that dataframe. 

You could rename it but why would you when you're just operating on the same dataframe. I could rename every single instance of the dataframe, e.g `df1, df2, ..., dfn` but 1) since a lot of Pandas operations are not in-place, I'd be wasting memory and 2) even more confusing to someone reading the code as then I start to think why would the engineer create N different copies of the dataframe if they didn't need them. If I just see one variable/dataframe thats being operated on, the intent is clear (even without documentation) that the code just does stuff to a single dataframe and then returns it.